tasks/libs/copyright.py

```diff
@@ -0,0 +1,184 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+
+import os
+import re
+import subprocess
+import sys
+from pathlib import Path, PurePosixPath
+
+GLOB_PATTERN = "**/*.go"
+
+COPYRIGHT_REGEX = [
+    r'^// Unless explicitly stated otherwise all files in this repository are licensed$',
+    r'^// under the Apache License Version 2.0\.$',
+    r'^// This product includes software developed at Datadog \(https://www\.[Dd]atadoghq\.com/\)\.$',
+    r'^// Copyright 20[1-3][0-9]-([Pp]resent|20[1-3][0-9]) Datadog, (Inc|Inmetrics)\.$',
+]
+
+# These path patterns are excluded from checks
+PATH_EXCLUSION_REGEX = [
+    '/third_party/golang/',
+    '/third_party/kubernetes/',
+]
+
+# These header matchers skip enforcement of the rules if found in the first
+# line of the file
+HEADER_EXCLUSION_REGEX = [
+    '^// Code generated ',
+    '^//go:generate ',
+    '^// Copyright.* OpenTelemetry Authors',
+    '^// Copyright.* The Go Authors',
+]
+
+
+COMPILED_COPYRIGHT_REGEX = [re.compile(regex, re.UNICODE) for regex in COPYRIGHT_REGEX]
+COMPILED_PATH_EXCLUSION_REGEX = [re.compile(regex, re.UNICODE) for regex in PATH_EXCLUSION_REGEX]
+COMPILED_HEADER_EXCLUSION_REGEX = [re.compile(regex, re.UNICODE) for regex in HEADER_EXCLUSION_REGEX]
+
+
+class CopyrightLinter:
+    """
+    This class is used to enforce copyright headers on specified file patterns
+    """
+
+    @staticmethod
+    def _get_repo_dir():
+        script_dir = PurePosixPath(__file__).parent
+        current_dir = Path.cwd()
+
+        os.chdir(script_dir)
+        repo_dir = subprocess.check_output(['git', 'rev-parse', '--show-toplevel']).decode(sys.stdout.encoding).strip()
+        os.chdir(current_dir)
+
+        return PurePosixPath(repo_dir)
+
+    @staticmethod
+    def _get_matching_files(root_dir, glob_pattern, exclude=None):
+        if exclude is None:
+            exclude = []
+
+        all_matching_files = Path(root_dir).glob(glob_pattern)
+
+        # `all_matching_files` is a generator so we have to do the counting ourselves
+        all_matching_files_cnt = 0
+
+        filtered_files = []
```

Comment:
nit: To me it would feel more 'Pythonic' to define an `is_excluded_file` predicate and then do
```python
filtered_files = [filepath for filepath in all_matching_files if is_excluded_file(filepath)]
```

Reply:
@mx-psi I made the inside loop be predicate-based but due to the generator being the iterable over which we evaluate this block and the fact that we need the total count for debugging, I left the outer loop as-is. I know the generator can be turned into a list easy for the length calculation but I think at that point we would be skyrocketing in mem usage so I opted against it.

Reply:
Makes sense, I think the predicate improvement makes this look better just by itself :)